lib/shopify_api/rest.ex

```diff
@@ -11,25 +11,76 @@ defmodule ShopifyAPI.REST do
   alias ShopifyAPI.JSONSerializer
   alias ShopifyAPI.REST.Request
 
-  @doc false
-  def get(%AuthToken{} = auth, path, params \\ []) do
-    Request.perform(auth, :get, path, "", params)
+  @doc """
+  Underlying utility retrieval function. The options passed affect both the
+  return value and, ultimately, the number of requests made to Shopify.
+
+  ## Options
+
+    `:pagination` - Can be `:none`, `:stream`, or `:auto`. Defaults to :auto
+    `:none` will only return the first page. You won't have access to the headers to manually
+      paginate.
+    `:auto` will block until all the pages have been retrieved and concatenated together.
+    `:stream` will return a `Stream`, prepopulated with the first page.
+  """
+  @spec get(AuthToken.t(), path :: String.t(), keyword(), keyword()) ::
+          {:ok, %{required(String.t()) => [map()]}} | Enumerable.t()
+  def get(%AuthToken{} = auth, path, params \\ [], options \\ []) do
+    collect_results = fn
+      %{} = result, {:ok, acc} -> {:cont, {:ok, [result | acc]}}
+      error, {:ok, _acc} -> {:halt, error}
+    end
+
+    case pagination(options) do
+      :none ->
+        with {:ok, response} <- Request.perform(auth, :get, path, "", params) do
+          {:ok, fetch_body(response)}
+        end
+
+      :stream ->
+        Request.stream(auth, path, params)
+
+      :auto ->
+        auth
+        |> Request.stream(path, params)
+        |> Enum.reduce_while({:ok, []}, collect_results)
+        |> case do
+          {:ok, results} -> {:ok, Enum.reverse(results)}
+          error -> error
+        end
```

Comment:
🛠 This additional collection reducer (and `collect_results` itself) could potentially be extracted to a helper, making this implementation easier to read.

```elixir
:auto ->
  auth |> Request.stream(path, params) |> collect_results()
```

Reply:
Something like this?
```elixir
@spec collect_results(Enumerable.t()) ::
          {:ok, list()} | {:error, HTTPoison.Response.t() | any()}
  defp collect_results(stream) do
    result_reducer = fn
      %{} = result, {:ok, acc} -> {:cont, {:ok, [result | acc]}}
      error, {:ok, _acc} -> {:halt, error}
    end

    stream
    |> Enum.reduce_while({:ok, []}, result_reducer)
    |> case do
      {:ok, results} -> {:ok, Enum.reverse(results)}
      error -> error
    end
  end
```

Reply:
Yep. I also think its less necessary to pull out the reducer function in this version, can just `Enum.reduce_while` and pipe that into `case`.